mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -1740,3 +1741,294 @@ Value mlir::rock::addPassThroughIndices(OpBuilder &b, Value transformed,
+    const bool mustBeModified = !mustBePreserved && !mustBeCompletelyRemoved;
+
+    // compute which lower dimensions must be preserved
+    switch (args.type) {
+    case TransformType::AddDim:
+    case TransformType::ConstDim:
+    case TransformType::Broadcast:
+    case TransformType::Merge: {
+      assert(!mustBeModified && "assume always one-to-many relation");
+      [[fallthrough]];
+    }
+    case TransformType::Unmerge: {
+      if (mustBePreserved || mustBeModified) {
+        // preserve all original lower dims
+        llvm::copy(tr.getLowerDims(), std::back_inserter(preservedLowerDims));
+      }
+      break;
+    }
+    case TransformType::PassThrough: {
+      assert(tr.getUpperDims().size() == tr.getLowerDims().size());
+      auto localLowerDims = tr.getLowerDims();
+
+      // propagate all preserved upper dims to lower dims
+      for (auto [idx, dim] : llvm::enumerate(tr.getUpperDims())) {
+        if (llvm::is_contained(preservedUpperDims, dim)) {
+          preservedLowerDims.push_back(localLowerDims[idx]);
+        }
+      }
+      break;
+    }
+    default:
+      return failure();
+    }
+
+    propagatePreservedNames<DimType::Upper>(tr, args);
+    propagatePreservedNames<DimType::Lower>(tr, args);
+
+    // re-compute transformation parameters
+    if (mustBePreserved || mustBeModified) {
+      switch (args.type) {
+      case TransformType::Unmerge: {
+        uint32_t total = 1;
+        for (auto globalDimIdx : preservedUpperDims) {
+          auto dimSize = origUpperBounds[globalDimIdx];
+          total *= dimSize;
+          args.params.push_back(dimSize);
+        }
+        assert(preservedLowerDims.size() == 1);
+        origLowerBounds[preservedLowerDims[0]] = total;
+        break;
+      }
+      case TransformType::PassThrough: {
+        // propagate possibly modified dimensions
+        for (auto [idx, upperDim] : llvm::enumerate(tr.getUpperDims())) {
+          const auto lowerDim = tr.getLowerDims()[idx];
+          origLowerBounds[lowerDim] = origUpperBounds[upperDim];
+        }
+        [[fallthrough]];
+      }
+      default: {
+        llvm::copy(tr.getParams(), std::back_inserter(args.params));
+        break;
+      }
+      }
+      argsVector.push_back(args);
+    }
+
+    std::get<DimType::Upper>(removedDims)
+        .append(getDifference<DimType::Upper>(tr, args));
+    std::get<DimType::Lower>(removedDims)
+        .append(getDifference<DimType::Lower>(tr, args));
+  }
+
+  // todo: use vector instead of set
+  // update remove indices set
+  for (auto dim : std::get<DimType::Lower>(removedDims)) {
+    newRemoveIndicesSet.insert(dim);
+  }
+
+  remapDims<DimType::Upper>(argsVector, removedDims);
+  remapDims<DimType::Lower>(argsVector, removedDims);
+
+  // build new transformations based on the computer preserved data
+  for (auto &args : argsVector) {
+    auto newTr =
+        TransformAttr::get(b.getContext(), args.type, args.params,
```

Comment:
Having understood the code upto this point,

It feels like we only need to maintain `preservedDims` and `params`.
All the other args could be filled late -- maybe using TransformAttr::get(b.context(), params, preservedDims, tr).

I d also say lets make a constructor like that in the tablegen. Then we dont need to maintain this TransformAttrArgs struct or we can still maintain that struct but only have  `preservedDims` and `params` in it.

Reply:
> Having understood the code upto this point,
> 
> It feels like we only need to maintain `preservedDims` and `params`. All the other args could be filled late -- maybe using TransformAttr::get(b.context(), params, preservedDims, tr).
> 
> I d also say lets make a constructor like that in the tablegen. Then we dont need to maintain this TransformAttrArgs struct or we can still maintain that struct but only have `preservedDims` and `params` in it.

You also need `rock::TransformType` to build transforms. In theory, you could build `preservedNames` later but you would need a reference to the original transform. It it easy to do everything in the first for-loop.

So, the reason why one needs to delay the re-building processes is because of dimensions re-numbering. One needs to aggregate the whole information about the removees and only then to make a call to `remapDims`